src/Settings/SettingsServiceProvider.php

```diff
@@ -41,4 +48,42 @@ public function register()
 
         $this->container->alias(SettingsRepositoryInterface::class, 'flarum.settings');
     }
+
+
+    public function boot(Container $container, Dispatcher $events, SettingsValidator $settingsValidator)
+    {
+        $events->listen(
+            Saved::class,
+            function (Saved $event) use ($container) {
+                $recompile = new RecompileFrontendAssets(
+                    $container->make('flarum.assets.forum'),
+                    $container->make(LocaleManager::class)
+                );
+                $recompile->whenSettingsSaved($event);
+
+                $validator = new ValidateCustomLess(
+                    $container->make('flarum.assets.forum'),
+                    $container->make('flarum.locales'),
+                    $container,
+                    $container->make('flarum.less.config')
+                );
+                $validator->whenSettingsSaved($event);
+            }
+        );
+
+        $events->listen(
+            Saving::class,
+            function (Saving $event) use ($container, $settingsValidator) {
+                $settingsValidator->assertValid($event->settings);
+
+                $validator = new ValidateCustomLess(
+                    $container->make('flarum.assets.forum'),
+                    $container->make('flarum.locales'),
+                    $container,
+                    $container->make('flarum.less.config')
+                );
+                $validator->whenSettingsSaving($event);
+            }
+        );
+    }
```

Comment:
The logic for validating custom less and for recompiling frontend assets should remain in `ForumServiceProvider`, as it's specific to the forum. The only event listening in this file should be running the settings validator on saved settings.

Reply:
OK, I misunderstood your original comment.